### PR TITLE
Implement proper ACL checks and suid permission

### PR DESCRIPTION
**Pull request by pkit**
Placeholder, not finished yet.

Tests needed still:

Anonymous (api/1.0):
- simple denial case
- if we reference two containers, 1 has setuid, 1 does not (api container has, other doesn't, and vice versa) <--- should be denied
- if setuid and read, should be allowed
- setuid on api container, and setuid and no acl on another container, check if we can write to other container (that is, create objects there)
- if job has read permissions and no setuid, anonymous should not work in this case (authorized users would be allowed, though-- because other user will be billed instead of the owner-- and we can't bill anonymous users, sadly =P )

### CODE REVIEW

**Comment by pkit** [comment]
Needs the following: 
- tests for anonymous access
- tests for "/open" for zapp (has one)
- tests for "/api"
- tests for "/open" an object with specific "content_type"

File: zerocloud/proxyquery.py [hunk 1]
```diff
@@ -950,17 +934,21 @@ def post_job(self, req):
...
             source_resp = source_req.get_response(self.app)
             if not is_success(source_resp.status_int):
                 return source_resp
-            del sink_req.headers['X-Zerovm-Source']
             sink_req.content_length = source_resp.content_length
             sink_req.content_type = source_resp.headers['Content-Type']
             sink_req.etag = source_resp.etag
             req_iter = iter(source_resp.app_iter)
             req = sink_req
+        self.authorize_job(req, remove_auth=False)
```
Comment (larsbutler) [inline zerocloud/proxyquery.py:951]: Shouldn't there be a `acl=` argument here? Perhaps: `self.authorize_job(req, acl='read_acl', remove_auth=False)`?

Comment (larsbutler) [inline zerocloud/proxyquery.py:951]: @pkit ^

Comment (pkit) [inline zerocloud/proxyquery.py:951]: No, job can be pre-authorized by owner only. No read or write permissions matter here.

Comment (larsbutler) [inline zerocloud/proxyquery.py:951]: Okay. By the way, I'm unable to get anonymous access working. Have you exercised this functionality yet?

File: test/unit/test_proxyquery.py [hunk 2]
```diff
@@ -3594,563 +3941,849 @@ def test_job_chain(self):
...
+        self.users = {'a': 'user', 'a1': 'user1'}
+
+    def tearDown(self):
+        if self.zerovm_mock:
+            os.unlink(self.zerovm_mock)
+
+    def remove_acls(self, url):
+        req = Request.blank(url,
+                            environ={'REQUEST_METHOD': 'POST'},
+                            headers={'X-Remove-Container-Read': 't',
```
Comment (larsbutler) [inline test/unit/test_proxyquery.py:4238]: What does the value `'t'` indicate? I don't understand the meaning.

Comment (pkit) [inline test/unit/test_proxyquery.py:4238]: "true", it really doesn't matter what you put in a value, but it cannot be empty string, because it's "undefined behavior" then :)

File: zerocloud/proxyquery.py (hunk 1 above)
Comment (pkit) [inline zerocloud/proxyquery.py:951]: I did not do any anonymous stuff tests yet, and there should be bugs there.

File: test/unit/test_proxyquery.py (hunk 2 above)
Comment (larsbutler) [inline test/unit/test_proxyquery.py:4238]: Okay, thanks!

**Comment by larsbutler** [comment]
@pkit Can you give me some ideas about how to set up the tests for `/api` and `/open`? I've been trying for several hours and can't make any sense of it.

**Comment by larsbutler** [comment]
For example, I can't figure out how--with the test utils and fixtures--to set up a zapp to handle requests for  `/open/account/container/foo.zapp`, or set the suid, or anything. I'm lost.

File: zerocloud/chain.py [hunk 3]
```diff
@@ -45,6 +45,8 @@ def handle_chain(self, env, start_response):
             new_req.environ['chain.input'] = env['chain.input']
             new_req.environ['chain.input_size'] = env['chain.input_size']
             new_req.environ['chain.input_type'] = env['chain.input_type']
+            if 'zerovm.source' in env:
```
Comment (larsbutler) [inline zerocloud/chain.py:48]: Why is this being copied? There's no explanation in comments or in a commit message.

File: zerocloud/configparser.py [hunk 4]
```diff
@@ -766,7 +766,7 @@ def resolve_path_info(self, account_name, replica_resolver):
                     node.access = 'PUT'
                 else:
                     node.path_info = default_path_info
-            if not top_channel:
+            else:
```
Comment (larsbutler) [inline zerocloud/configparser.py:769]: Why is this change necessary?

**Comment by larsbutler** [comment]
According to the [spec](https://github.com/zerovm/zerocloud/issues/144) for this feature, the suid stuff should only apply to api/1.0 and open/1.0 features. However, the tests so far implemented on this branch test suid features of everything BUT the REST features.

Why is that? Do we need to update the spec?

File: zerocloud/chain.py (hunk 3 above)
Comment (pkit) [inline zerocloud/chain.py:48]: We need to know which chains come from api/1.0 requests and which are just user-created. Only api/1.0 ones need to be checked on suid secirity.

File: zerocloud/configparser.py (hunk 4 above)
Comment (pkit) [inline zerocloud/configparser.py:769]: Otherwise it leaves `node.path_info = None` which is not desired (more checks down the line).

**Comment by pkit** [comment]
Added example with API/1.0 call, fixed a bug regarding anonymous access to API endpoint.

File: zerocloud/chain.py (hunk 3 above)
Comment (larsbutler) [inline zerocloud/chain.py:48]: Okay, can you please add a comment explaining that here?

**Comment by larsbutler** [comment]
See https://github.com/zerovm/zerocloud/pull/172 for a rebased version of this branch.
